### DISCUSSION TITLE
when should we create and clear interaction handle in the drawer navigator

### DISCUSSION THREAD

**Discussion by ntdiary** [react-navigation/react-navigation#10899] [General]
**drawer v6.5.0**

**we want to defer a task until the drawer animation end**,
but there seems to be some problems here, and I have done some tests:

this is a commit for test: https://github.com/ntdiary/react-navigation/commit/a1f504ed1964107e4736818a9804cecae51bccff

In this commit, console will print labels and timestamps at following moments:
- invoke `createInteractionHandle`, print `create handle`
- invoke `clearInteractionHandle`, print `clear  handle`
- drawer spring animation end, print `animation end`
- the deferred task is executed, print `deferred task`
# screenshot
- run or reload example app, clear the initialization log
- open the drawer and check the log

### legacy mode
<img width="1572" alt="ios-legacy" src="https://user-images.githubusercontent.com/8579651/193386408-45d45b5d-9d17-469c-868f-82f2b75550f0.png">

- `create handle` and `clear  handle` are printed multiple times
- `deferred task` is printed after the first `clear  handle` instead of `animation end`

### modern mode
<img width="1498" alt="ios-modern" src="https://user-images.githubusercontent.com/8579651/193386463-4512498e-1153-4481-b3de-3944137fbf22.png">

- `deferred task` is printed after `clear  handle` instead of `animation end`

## web (emulated device)
The current implementation of `react-native-web` stubs out `createInteractionHandle` and `clearInteractionHandle`.
And `runAfterInteractions` will [call the task immediately](https://github.com/necolas/react-native-web/blob/dee258ae02188befca4192ea502d4fbebc478341/packages/react-native-web/src/exports/InteractionManager/index.js#L35) instead of putting it in the [queue](https://github.com/facebook/react-native/blob/43636267011a97490ed7495b08e500c5d0d54872/Libraries/Interaction/InteractionManager.js#L101) like `react-native`.
So the `deferred task` is always printed immediately.
(I posted a feature request [here](https://github.com/necolas/react-native-web/issues/2399))
### legacy mode
<img width="1335" alt="web-legacy" src="https://user-images.githubusercontent.com/8579651/193386560-94a29181-4e99-4be0-a145-4567c2320727.png">

- `deferred task` is printed before `animation end`

### modern mode
<img width="1354" alt="web-modern" src="https://user-images.githubusercontent.com/8579651/193386564-893f24c2-a046-4063-aa08-8fe9f849f03f.png">

- `deferred task` is printed before `animation end`
- `create handle` and `clear handle` are not printed

# expect
Could we change when to create and clear interaction handle?
### legacy mode
- Clear interaction handle only after spring animation ends.
### modern mode
- Check if there is an interaction handle before the spring animation starts, and create it if not
- Clear interaction handle only after spring animation ends.

**Comment by satya164** [comment]
It's OK to create and clear multiple interaction handles between the gesture start and gesture end, but ideally there should be at least one handle present from start of the gesture to the end of the animation when triggered via gesture and the start of the animation to end of animation when triggered programmatically. PR is welcome to make it work like this if it's currently buggy.

The way it's implemented in React Native Web is unfortunate, so I guess we'd also need to implement same events as stack:
- `gestureStart`
- `gestureEnd`
- `gestureCancel`
- `transitionStart`
- `transitionEnd`

**Reply by ntdiary** [reply]
Hi, sorry for the long delay, I submitted a [PR](https://github.com/necolas/react-native-web/pull/2425) to the react-native-web first, and it's under review. 🙂

**Reply by satya164** [reply]
@ntdiary Thank you! Looking forward to it getting merged.
